adaptdl/adaptdl/goodput_test.py

```diff
@@ -0,0 +1,224 @@
+                                       atomic_bsz_range=(64, 256))
+    assert(bsz == 128), "expected bsz = 128, got {}".format(bsz)
+    assert(isinstance(goodput, float))
+
+    replicas = np.asarray(range(1, 20))
+    # single-node
+    goodput, bsz, steps = fun.optimize(np.ones_like(replicas), replicas,
+                                       max_batch_size=1280,
+                                       atomic_bsz_range=(64, 256))
+    assert(np.all(np.logical_or(bsz >= np.ceil(128 / replicas).astype(int),
+                                goodput == 0.0)))
+    assert(np.all(np.logical_or(bsz >= (64),
+                                goodput == 0.0)))
+    assert(np.all(bsz <= (256)))
+    assert(np.all(np.logical_or(bsz * replicas <= 1280 + replicas,
+                                goodput == 0.0)))
+    assert(bsz[0] == 128)
+    assert(np.all(steps == 0))
+    # multi-node
+    goodput, bsz, steps = fun.optimize(replicas, replicas,
+                                       max_batch_size=1280,
+                                       atomic_bsz_range=(64, 256))
+    assert(np.all(np.logical_or(bsz >= np.ceil(128 / replicas).astype(int),
+                                goodput == 0.0)))
+    assert(np.all(np.logical_or(bsz >= (64),
+                                goodput == 0.0)))
+    assert(np.all(bsz <= (256)))
+    assert(np.all(np.logical_or(bsz * replicas <= 1280 + replicas,
+                                goodput == 0.0)))
+    assert(bsz[0] == 128)
+    assert(np.all(steps == 0))
+
+
+@pytest.mark.parametrize("perf_params", PERF_PARAMS)
+@pytest.mark.parametrize("grad_params", GRAD_PARAMS)
+def test_optimize_accumulation(perf_params, grad_params):
+    grad_params = {"norm": grad_params[0], "var": grad_params[1]}
+    fun = GoodputFunction(perf_params, grad_params, 128)
+    goodput, bsz, steps = fun.optimize(1, 1, max_batch_size=1280,
+                                       atomic_bsz_range=(64, 256),
+                                       accumulation=True)
+    assert(isinstance(goodput, float))
+
+    replicas = np.asarray(range(1, 20))
+    # single-node
+    goodput, bsz, steps = fun.optimize(np.ones_like(replicas), replicas,
+                                       max_batch_size=1280,
+                                       atomic_bsz_range=(64, 256),
+                                       accumulation=True)
+    assert(np.all(np.logical_or(bsz >= np.ceil(128 / replicas).astype(int),
+                                goodput == 0.0)))
+    assert(np.all(np.logical_or(bsz >= (64),
+                                goodput == 0.0)))
+    assert(np.all(bsz <= (256)))
+    assert(np.all(np.logical_or(bsz * replicas * (steps + 1) <
+                                1280 + replicas * (steps + 1),
+                                goodput == 0.0)))
+    assert(np.all(steps <= 15))
+    assert(np.all(steps >= 0))
+    assert(np.all(np.logical_or(np.multiply(steps, bsz) >= 256,
```

Comment:
I dont think this invariant is maintained anymore. I think it should be `np.logical_or(replicas > 1, np.logical_or(bsz == 128, steps > 0))`